server/service/microsoft_mdm.go

```diff
@@ -572,6 +784,234 @@ func (svc *Service) GetMDMWindowsPolicyResponse(ctx context.Context, authToken s
+	if len(authToken) == 0 {
+		return nil, fleet.NewInvalidArgumentError("enroll response", "authToken is empty")
+	}
+
+	// Validate the binary security token
+	err := svc.validateBinarySecurityToken(ctx, authToken)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "validate binary security token")
+	}
+
+	// Checking if this device is already MDM enrolled
+	err = svc.isWindowsDeviceAlreadyMDMEnrolled(ctx, secTokenMsg)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "device is already enrolled")
+	}
+
+	// Getting the the device provisioning information in the form of a WapProvisioningDoc
+	deviceProvisioning, err := svc.getDeviceProvisioningInformation(ctx, secTokenMsg)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "device is already enrolled")
+	}
+
+	// Token is authorized
+	svc.authz.SkipAuthorization(ctx)
+
+	// Getting the RequestSecurityTokenResponseCollection message content
+	secTokenResponseCollectionMsg, err := NewRequestSecurityTokenResponseCollection(deviceProvisioning)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "creation of RequestSecurityTokenResponseCollection message")
+	}
+
+	// RequestSecurityTokenResponseCollection message is ready
+	// The identity and provisioning information will be sent to the Windows MDM Enrollment Client
+
+	// But before doing that, let's save the device information to the list of MDM enrolled MDM devices
+	err = svc.storeWindowsMDMEnrolledDevice(ctx, secTokenMsg)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "enrolled device information cannot be stored")
+	}
+
+	return &secTokenResponseCollectionMsg, nil
+}
+
+// isWindowsDeviceAlreadyMDMEnrolled checks if the device is already enrolled by checking the device
+// UUID against the list of enrolled devices
+func (svc *Service) isWindowsDeviceAlreadyMDMEnrolled(ctx context.Context, secTokenMsg *fleet.RequestSecurityToken) error {
+	// Getting the DeviceID from the RequestSecurityToken msg
+	reqDeviceID, err := GetContextItem(secTokenMsg, mdm.ReqSecTokenContextItemDeviceID)
+	if err != nil {
+		return err
+	}
+
+	// Checking the storage to see if the device is already enrolled
+	if _, err := svc.ds.MDMWindowsGetEnrolledDevice(ctx, reqDeviceID); err != nil {
+		if fleet.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	return nil
+}
+
+// getDeviceProvisioningInformation returns a valid WapProvisioningDoc
+// This is the provisioning information that will be sent to the Windows MDM Enrollment Client
+// This information is used to configure the device management client
+// See section 2.2.9.1 for more details on the XML provision schema used here
+// https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-mde2/35e1aca6-1b8a-48ba-bbc0-23af5d46907a
+func (svc *Service) getDeviceProvisioningInformation(ctx context.Context, secTokenMsg *fleet.RequestSecurityToken) (string, error) {
+	const (
+		error_tag = "wap provisioning: "
```

Comment:
remember to use `mixedCaps` for variable names.

nit: I would argue that if the error prefix is the same, then the caller should be adding it.

generally, I try to add more context to the particular failure instead of a generic prefix.

Reply:
That's a good point, I've just modified the code to no wrap `err`